### PR TITLE
Add DeleteLine and CutLine operations to interactive window

### PR DESCRIPTION
**Pull request by genlu**
Works with both empty and non-empty selection. The goal is to be consistent with `Delete` and `Cut ` behavioral wise.

This should address https://github.com/dotnet/roslyn/issues/5309 and https://github.com/dotnet/roslyn/issues/5462

@amcasey @ManishJayaswal @cston @tmat @KevinH-MS

### CODE REVIEW

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 1]
```diff
@@ -2340,8 +2317,90 @@ public void Cut()
...
+                    selection.AnchorPoint.GetLineAndColumn(out startLine, out startCol);
+                    selection.ActivePoint.GetLineAndColumn(out endLine, out endCol);
+
+                    if (startLine.LineNumber > endLine.LineNumber)
+                    {
+                        var temp = startLine;
+                        startLine = endLine;
+                        endLine = temp;
+                    }
+
```
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: What happens when startCol > endCol but startLine is same as endLine? Do we need to swap the startCol and endCol in that case?

Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Could we use selection.Start and selection.End instead? Start &lt;= End

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 2]
```diff
@@ -2340,8 +2317,90 @@ public void Cut()
...
+                        CutOrDeleteCurrentLine(isCut); 
+                        Home(false);                
+                    }
+                }
+                else
+                {
+                    var selection = TextView.Selection;
+
+                    var anchor = selection.AnchorPoint;
+                    var active = selection.ActivePoint;
```
Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Unused.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 3]
```diff
@@ -2361,16 +2420,20 @@ private void CutOrDeleteCurrentLine(bool isCut)
             /// </summary>                 
             private bool CutOrDeleteSelection(bool isCut)
             {
-                if (!TextView.Selection.IsEmpty)
+                var selection = TextView.Selection;
+                if (!selection.IsEmpty)
                 {
-                    bool isEditable = TextView.Selection.Mode == TextSelectionMode.Stream ?
-                                        IsStreamSelectionInEditableBuffer(TextView.Selection) :
+                    // Even though `IsSelectionInEditableBuffer` and `CutOrDelete` is sufficient to
```
Comment (cston) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Even though OverlapsWithEditableBuffer and ...?

Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: @cston Good catch! forgot to update comment.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 1 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: @ManishJayaswal @cston We don't use values of columns here. Anyway, I will rewrite this method as Chuck suggested, which will get rid of unused column as well as the swap altogether.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 4]
```diff
@@ -1295,6 +1295,215 @@ public void PasteWithSelectionInReadonlyArea()
...
+");                                                                                                                        
+            var caret = Window.TextView.Caret;                               
+
+            // DeleteLine with caret in readonly area
+            caret.MoveToPreviousCaretPosition();
+            caret.MoveToPreviousCaretPosition();
+            caret.MoveToPreviousCaretPosition();
+
+            AssertCaretVirtualPosition(1, 1);
+            Task.Run(() => Window.Operations.DeleteLine()).PumpingWait();
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Does this have to happen on a separate thread?

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 5]
```diff
@@ -1295,6 +1295,215 @@ public void PasteWithSelectionInReadonlyArea()
...
+            Assert.Equal("> 1\r\n1\r\n> ", Window.TextView.TextBuffer.CurrentSnapshot.GetText());
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: If there isn't already a helper method in this class for ```Window.TextView.TextBuffer.CurrentSnapshot.GetText());```, then there should be.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 6]
```diff
@@ -1295,6 +1295,215 @@ public void PasteWithSelectionInReadonlyArea()
...
+            AssertCaretVirtualPosition(1, 1);
+
+            // DeleteLine with caret in active prompt
+            caret.MoveToNextCaretPosition();
+            caret.MoveToNextCaretPosition();
+            caret.MoveToNextCaretPosition();
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: After each group of caret moves, can you please add a comment describing where we ended up?

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 7]
```diff
@@ -1295,6 +1295,215 @@ public void PasteWithSelectionInReadonlyArea()
...
+            Window.InsertCode("int x");
+            Window.Operations.BreakLine();
+            Window.InsertCode(";");
+            for (int i = 0; i < 11; ++i)
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: What's 11?  If it's the buffer length, can we just use that?

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 8]
```diff
@@ -1295,6 +1295,215 @@ public void PasteWithSelectionInReadonlyArea()
...
+            Window.Operations.SelectAll();
+            Window.Operations.SelectAll();
+            Task.Run(() => Window.Operations.DeleteLine()).PumpingWait();
+            Assert.Equal("> 1\r\n1\r\n> ", Window.TextView.TextBuffer.CurrentSnapshot.GetText());
+            AssertCaretVirtualPosition(2, 2);
+            Assert.True(selection.IsEmpty);
+        }
+
+        [Fact]
+        public void CutineWithOutSelection()
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Missing "L".

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 9]
```diff
@@ -2361,16 +2403,20 @@ private void CutOrDeleteCurrentLine(bool isCut)
...
-                if (!TextView.Selection.IsEmpty)
+                var selection = TextView.Selection;
+                if (!selection.IsEmpty)
                 {
-                    bool isEditable = TextView.Selection.Mode == TextSelectionMode.Stream ?
-                                        IsStreamSelectionInEditableBuffer(TextView.Selection) :
+                    // Even though `OverlapsWithEditableBuffer` and `CutOrDelete` is sufficient to 
+                    // delete editable selection,  we still need to handle box selection 
+                    // differently to move caret to appropiate location after deletion.
+                    bool isEditable = selection.Mode == TextSelectionMode.Stream ?
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: In C#, the usual formatting is 
```C#
cond
    ? true_expr
    : false_expr;
```

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 10]
```diff
@@ -2340,8 +2317,73 @@ public void Cut()
...
+                {
                     MoveCaretToClosestEditableBuffer();
                 }
+                return handled;
+            }
+
+            /// <summary>Implements <see cref="IInteractiveWindowOperations2.DeleteLine"/>.</summary>
+            public void DeleteLine()
+            {
+                _historySearch = null;
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Push this down into ```CutOrDeleteLine```?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 11]
```diff
@@ -2340,8 +2317,73 @@ public void Cut()
...
+            {
+                if (TextView.Selection.IsEmpty)
+                {
+                    var caret = TextView.Caret;
+                    var position = caret.Position.BufferPosition;
+                    if (MapToEditableBuffer(position) != null ||
+                        IsInActivePrompt(position))
+                    {
+                        CutOrDeleteCurrentLine(isCut); 
+                        Home(false);
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Named argument?

File: src/InteractiveWindow/Editor/InteractiveWindow.cs [hunk 12]
```diff
@@ -376,8 +376,17 @@ bool IInteractiveWindowOperations.BreakLine()
         bool IInteractiveWindowOperations.Return()
         {
             return UIThread(uiOnly => uiOnly.Return());
+        }
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.cs]: Whitespace change?

File: src/InteractiveWindow/Editor/IInteractiveWindowOperations2.cs [hunk 13]
```diff
@@ -8,5 +8,15 @@ public interface IInteractiveWindowOperations2 : IInteractiveWindowOperations
         /// Copies the current selection to the clipboard.
         /// </summary>
         void Copy();
+
+        /// <summary>
+        /// Delete Line; Delete all selected lines, or the current line if no selection.  
+        /// </summary>
+        void DeleteLine();
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/IInteractiveWindowOperations2.cs]: As we discussed offline, there's no ```CopyLine``` since there's no ```Edit.LineCopy``` in VS (whereas ```Edit.LineDelete``` and ```Edit.LineCut``` do exist).

**Comment by amcasey** [comment]
:+1:

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 7 above)
Comment (genlu) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: It's the position in the middle of the first active prompt. I used a loop just to avoid a list of call to `MoveToNextCaretPosition`.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 5 above)
Comment (genlu) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: okey dokey

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 4 above)
Comment (genlu) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: This is something I don't understand actually. What operation need to happen on a separate thread?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 9 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Fixed, thanks!

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 10 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: I prefer to keep it here since it has nothing to do with the logic of helper method.

File: src/InteractiveWindow/Editor/InteractiveWindow.cs (hunk 12 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.cs]: Somehow Git (or whatever diff tool used by GitHub) decides this is the diff. But I did miss a newline before the next `#endregion`.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 6 above)
Comment (genlu) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Sure! It will be in another PR though, since I'd like to add comments for tests I added before as well.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 4 above)
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Generally speaking, there are two reasons we might invoke a method on a background thread:
 1. We're testing how it behaves on a background thread.  Those tests have already been written, but feel free to expand them.
 2. It should have been async, but it isn't and we have to use ```PumpingWait``` to determine when it finishes.  We should never need to do this for new methods, because they can be made async.

Comment (genlu) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Then you are right, no need for a separate thread. Will fix.

**Comment by genlu** [comment]
ping @ManishJayaswal @cston @tmat @KevinH-MS

**Comment by genlu** [comment]
@dotnet-bot Retest this, please

**Comment by genlu** [comment]
@dotnet-bot Retest this, please

**Comment by genlu** [comment]
test Linux please
